### PR TITLE
PLINQ parallel processing for VoiceAttack extended variables

### PR DESCRIPTION
**Pull request by Tkael**
Ref. similar PR #1706.

### CODE REVIEW

**Comment by Tkael** [comment]
Placed on hold while I evaluate performance impact.

**Comment by Tkael** [comment]
Performance improved by up to 500 ms when writing variables from a ShipLoadoutEvent 20 times.

File: Tests/VoiceAttackPluginTests.cs [hunk 1]
```diff
@@ -24,27 +24,62 @@ public void WriteToLog(string msg, string color = null)
 
         public void SetText(string varName, string value)
         {
-            vaVars.Add(varName, value);
+            if (vaVars.ContainsKey(varName))
```
Comment (richardbuckle) [inline Tests/VoiceAttackPluginTests.cs]: Are we sure this is necessary? Dictionaries are highly optimized and this looks like paying the cost for the `ContainsKey()` twice: once by us and once by the dictionary. 
I'd go with just `vaVars[varName] = value;` unless I had strong evidence that this is  big win.

File: VoiceAttackResponder/VoiceAttackVariables.cs [hunk 2]
```diff
@@ -89,149 +89,152 @@ public static void setEventValues(dynamic vaProxy, Event theEvent, List<string>
...
         /// </summary>
-        public static void setEventExtendedValues(dynamic vaProxy, string prefix, dynamic json, List<string> setKeys)
+        public static void setEventExtendedValuesParallel(dynamic vaProxy, string prefix, dynamic json, List<string> setKeys)
         {
-            foreach (JProperty child in json)
+            var jProperties = ((JObject)json).Properties();
+            jProperties.AsParallel().Select(child => setEventExtendedValue(vaProxy, prefix, child, ref setKeys));
+        }
+
+        public static void setEventExtendedValue(dynamic vaProxy, string prefix, JProperty child, ref List<string> setKeys)
```
Comment (richardbuckle) [inline VoiceAttackResponder/VoiceAttackVariables.cs]: This is problematic as `vaProxy` is almost certainly not thread-safe.

File: VoiceAttackResponder/VoiceAttackVariables.cs [hunk 3]
```diff
@@ -89,149 +89,152 @@ public static void setEventValues(dynamic vaProxy, Event theEvent, List<string>
...
-                    Logging.Debug("Ignoring key " + child.Name);
-                    continue;
-                }
-                // We also ignore any keys that we have already set elsewhere
-                if (setKeys.Contains(child.Name))
-                {
-                    Logging.Debug("Skipping already-set key " + child.Name);
-                    continue;
-                }
+                Logging.Debug("Ignoring key " + child.Name);
```
Comment (richardbuckle) [inline VoiceAttackResponder/VoiceAttackVariables.cs]: All this `Logging.Debug()` is going to re-serialise us.

**Comment by richardbuckle** [comment]
Back on hold as I have some concerns.

File: Tests/VoiceAttackPluginTests.cs (hunk 1 above)
Comment (Tkael) [reply Tests/VoiceAttackPluginTests.cs]: Agreed.

**Comment by Tkael** [comment]
Fair enough. I'll withdraw and think about this one some more.

**Pull request closed without merging**